teuthology/provision/maas.py

```diff
@@ -0,0 +1,441 @@
+                )
+        elif raise_on_error:
+            raise RuntimeError(
+                "Failed to lock machine %s, system_id: %s",
+                self.shortname, self.system_id
+            )
+
+    def unlock_machine(self, raise_on_error=True):
+        """Unlock the machine
+
+        :param raise_on_error: Whether or not to raise an exception
+                               if the request is unsuccessful (default: True)
+        """
+        resp = self.do_request(
+            f"/machines/{self.system_id}/op-unlock",
+            method="POST",
+            raise_on_error=raise_on_error
+        )
+        if resp.json():
+            if resp.json().get("locked") and raise_on_error:
+                raise RuntimeError(
+                    "Machine %s locking failed, status: %s",
+                    self.shortname, resp.json().get("locked")
+                )
+        elif raise_on_error:
+            raise RuntimeError(
+                "Failed to lock machine %s, system_id: %s",
+                self.shortname, self.system_id
+            )
+
+    def deploy_machine(self):
+        """Deploy the machine"""
+        image_data = self.get_image_data()
+        if image_data.get("type").lower() != "synced":
+            raise RuntimeError(
+                "MaaS image %s is not synced, current status: %s",
+                image_data.get("name"), image_data.get("type")
+            )
+        log.info(
+            "Deploying machine %s with image %s",
+            self.shortname, image_data.get("name")
+        )
+    
+        files = {
+            "distro_series": (None, image_data.get("name")),
+            "user_data": (None, self._get_user_data()),
+        }
+        resp = self.do_request(
+            f"/machines/{self.system_id}/op-deploy", method="POST", files=files
+        ).json()
+        if not resp:
+            raise RuntimeError(
+                "Failed to deploy machine %s, system_id: %s",
+                self.shortname, self.system_id
+            )
+        if not resp.get("status_name").lower() == "deploying":
+            raise RuntimeError(
+                "Machine %s deployment failed, status: %s",
+                self.shortname, resp.get("status_name")
+            )
+
+    def release_machine(self, erase=True):
+        """Release the machine
+
+        params: Optional parameters for the erasing disks
+        """
+        data = { "erase": erase }
+        resp = self.do_request(
+            f"/machines/{self.system_id}/op-release", method="POST", data=data
+        ).json()
+        if not resp:
+            raise RuntimeError(
+                "Failed to release machine %s, system_id: %s",
+                self.shortname, self.system_id
+            )
+        if not resp.get("status_name").lower() == "disk erasing":
+            raise RuntimeError(
+                "Machine %s releasing failed, status: %s",
+                self.shortname, resp.get("status_name")
+            )
```

Comment:
As before, if these truly are two distinct conditions let's try to distinguish them more clearly in the error reporting.

Reply:
Same as above these two conditions are distinct but the first check should be taken care in do_request and hence not needed here. Updated code accordingly.